synapse/storage/devices.py

```diff
@@ -552,12 +617,25 @@ def store_device(self, user_id, device_id, initial_device_display_name):
                     "user_id": user_id,
                     "device_id": device_id,
                     "display_name": initial_device_display_name,
+                    "hidden": False,
                 },
                 desc="store_device",
                 or_ignore=True,
             )
+            if not inserted:
+                # if the device already exists, check if it's a real device, or
+                # if the device ID is reserved by something else
+                hidden = yield self._simple_select_one_onecol(
+                    "devices",
+                    keyvalues={"user_id": user_id, "device_id": device_id},
+                    retcol="hidden",
+                )
+                if hidden:
+                    raise StoreError(400, "The device ID is in use", Codes.FORBIDDEN)
             self.device_id_exists_cache.prefill(key, True)
             defer.returnValue(inserted)
+        except StoreError as e:
+            raise e
```

Comment:
You want to allow the existing exception to propagate, not rethrow it with a new stack trace:

```suggestion
            raise
```